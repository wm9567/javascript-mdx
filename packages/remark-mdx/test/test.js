const unified = require('unified')
const remarkParse = require('remark-parse')
const remarkStringify = require('remark-stringify')

const remarkMdx = require('..')
const mdxAstToMdxHast = require('../../mdx/mdx-ast-to-mdx-hast')
const mdxHastToJsx = require('../../mdx/mdx-hast-to-jsx')

const FIXTURE = `
import Foo from './bar'
export { Baz } from './foo'
export default Foo

# Hello, world! <Foo bar={{ baz: 'qux' }} />

<Baz>
  Hi!
</Baz>

<Paragraph bg='red.500' color='white'>Foo</Paragraph>

<Button>
  Hi!
</Button>

<>Foo</>

<>
  Foo
</>

<h1>Hello, world!</h1>
`

// Manually apply all mdx transformations for now
// so we don't introduce a circular dependency.
//
// When remark-mdx is used by core we can remove
// these integration tests
const transpile = mdx => {
  const result = unified()
    .use(remarkParse)
    .use(remarkMdx)
    .use(mdxAstToMdxHast)
    .use(mdxHastToJsx)
    .processSync(mdx)

  return result.contents
}

const parse = mdx => {
  const result = unified()
    .use(remarkParse)
    .use(remarkMdx)
    .parse(mdx)

  return result
}

const stringify = mdx => {
  const result = unified()
    .use(remarkParse)
    .use(remarkStringify)
    .use(remarkMdx)
    .processSync(mdx)

  return result.contents
}

it('correctly transpiles', () => {
  const result = transpile(FIXTURE)

  expect(result).toMatchSnapshot()
})

it('maintains the proper positional info', () => {
  const result = parse(FIXTURE)

  expect(result).toMatchSnapshot()
})

it('does not wrap an block level elements in a paragraph', () => {
  const result = transpile(FIXTURE)

  expect(result).not.toMatch(/<p><Baz/)
  expect(result).not.toMatch(/<p><Button/)
  expect(result).not.toMatch(/<p><Paragraph>/)
  expect(result).not.toMatch(/<p><>/)
  expect(result).not.toMatch(/<p><h1/)
})

it('removes newlines between imports and exports', () => {
  const fixture = [
    'import Foo1 from "./foo"',
    'import Foo2 from "./foo"',
    'import Foo3 from "./foo"',
    'export const foo = "bar"',
    'export default props => <article {...props} />',
    'export const fred = "flintstone"'
  ].join('\n')

  const result = stringify(fixture)

  expect(result).toMatchSnapshot()
})
