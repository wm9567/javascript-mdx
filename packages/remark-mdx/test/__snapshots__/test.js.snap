// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly transpiles 1`] = `
"import Foo from './bar'

export { Baz } from './foo'

const layoutProps = {
  
};
const MDXLayout = Foo
export default function MDXContent({ components, ...props }) {
  return (
    <MDXLayout
      {...layoutProps}
      {...props}
      components={components}>



<h1>{\`Hello, world! \`}<Foo bar={{ baz: 'qux' }} /></h1>
<Baz>
  Hi!
</Baz>
    </MDXLayout>
  )
}
MDXContent.isMDXComponent = true"
`;

exports[`maintains the proper positional info 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 1,
          "line": 3,
          "offset": 25,
        },
        "indent": Array [
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "import",
      "value": "import Foo from './bar'
",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 1,
          "line": 4,
          "offset": 53,
        },
        "indent": Array [
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 25,
        },
      },
      "type": "export",
      "value": "export { Baz } from './foo'
",
    },
    Object {
      "default": true,
      "position": Position {
        "end": Object {
          "column": 19,
          "line": 4,
          "offset": 71,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 4,
          "offset": 53,
        },
      },
      "type": "export",
      "value": "export default Foo",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 17,
              "line": 6,
              "offset": 89,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 6,
              "offset": 75,
            },
          },
          "type": "text",
          "value": "Hello, world! ",
        },
        Object {
          "position": Position {
            "end": Object {
              "column": 45,
              "line": 6,
              "offset": 117,
            },
            "indent": Array [],
            "start": Object {
              "column": 17,
              "line": 6,
              "offset": 89,
            },
          },
          "type": "jsx",
          "value": "<Foo bar={{ baz: 'qux' }} />",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 45,
          "line": 6,
          "offset": 117,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 6,
          "offset": 73,
        },
      },
      "type": "heading",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 7,
          "line": 10,
          "offset": 137,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 8,
          "offset": 119,
        },
      },
      "type": "jsx",
      "value": "<Baz>
  Hi!
</Baz>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 11,
      "offset": 138,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`removes newlines between imports and exports 1`] = `
"import Foo1 from \\"./foo\\"
import Foo2 from \\"./foo\\"
import Foo3 from \\"./foo\\"

export const foo = \\"bar\\"

export default props => <article {...props} />

export const fred = \\"flintstone\\"
"
`;
